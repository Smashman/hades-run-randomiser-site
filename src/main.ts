import { Run, RunOptions } from './run';

const runOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        heatRange: {
            min: 4,
            max: 12,
        },
    }
};

const run = Run.generateRandomRun(runOptions);

console.log('run', run);
console.log('run.weapon', run.weapon);
console.log('run.weaponAspect', run.weaponAspect);
console.log('run.keepsake', run.keepsake);
console.log('run.companion', run.companion);
console.log('run.mirror', run.mirror);
console.log('run.pact', run.pact);