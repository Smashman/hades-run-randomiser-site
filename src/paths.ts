export const imgPath = 'public/img';
export const unknownIconPath = `${imgPath}/misc/unknown_icon.png`;