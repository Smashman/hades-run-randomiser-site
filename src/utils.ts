export const getRandomItemFromArray = <T>(array: T[]) => array[Math.floor(Math.random() * array.length)];
export const fillArray = <T>(length: number, fillValue: T) => Array.from<T>({length: length}).fill(fillValue);

export interface Options {
    enabled?: boolean;
}

export abstract class Item {
    isUnlocked: boolean = false;

    unlock() {
        this.isUnlocked = true;
    }

    lock() {
        this.isUnlocked = false;
    }
}

export abstract class ItemList<T extends Item> {
    items: T[] = [];

    addItem(item: T) {
        this.items = [...this.items, item];
        return this;
    }

    addItems(items: T[]) {
        items.forEach(item => this.addItem(item));
        return this;
    }

    get unlocked() {
        return this.items.filter(item => item.isUnlocked);
    }

    unlockAll() {
        this.items.forEach(item => item.unlock());
    }

    getRandom(): T {
        return getRandomItemFromArray(this.unlocked) ?? null;
    }
}

export class Level {
    constructor(protected _value: number, readonly min: number, readonly max: number) {}

    set value(setTo: number) {
        if (setTo >= this.max) {
            this._value = this.max;
        }
        else if (setTo <= this.min) {
            this._value = this.min;
        }
        else {
            this._value = setTo;
        }
    }

    get value() {
        return this._value;
    }

    isMinLevel(): boolean {
        return this.value <= this.min;
    }

    isMaxLevel(): boolean {
        return this.value >= this.max;
    }

    setMaxLevel(): void {
        this.value = this.max;
    }

    addValue(change: number): number {
        this.value = this.value + change;
        return this.value;
    }

    increment(): number {
        return this.addValue(1);
    }

    decrement(): number {
        return this.addValue(-1);
    }
}