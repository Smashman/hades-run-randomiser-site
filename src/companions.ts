import { getRandomItemFromArray, Item, Level, Options } from './utils';
import { ItemList } from './utils';

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export class Companion extends Item {
    level: Level = new Level(1, Companion.minLevel, Companion.maxLevel);
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static minLevel: number = 1;
    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }

    toStorableData() {
        const itemData = super.toStorableData();
        return {
            ...itemData,
            level: this.level.value,
            isMaxCodex: this.isMaxCodex,
        }
    }
}