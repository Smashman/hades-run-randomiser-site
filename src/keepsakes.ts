import { getRandomItemFromArray, Options, ItemList, Item, Level, StorableItemListData, StorableItemData } from './utils';

export interface KeepsakeOptions extends Options {
    ignoreMaxLevel?: boolean;
    changeBetweenBiomes?: boolean;
}

export interface StorableKeepsakesData extends StorableItemListData<StorableKeepsakeData> {

}

export class Keepsakes extends ItemList<Keepsake> {
    getRandomKeepsakes(keepsakeOptions?: KeepsakeOptions): KeepsakeConfiguration {
        const keepsakes = (keepsakeOptions?.ignoreMaxLevel && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(keepsake => !keepsake.isMaxLevel)] : [...this.unlocked];

        if (keepsakes.length === 0) {
            return null;
        }

        let keepsakesChosen: Keepsake[] = [getRandomItemFromArray(keepsakes)];

        if (keepsakeOptions?.changeBetweenBiomes) {
            for (let i = 1; i <= 3; i++) {
                keepsakesChosen = [...keepsakesChosen, getRandomItemFromArray(keepsakes.filter(keepsake => !keepsakesChosen.includes(keepsake))) ?? null];
            }
        }

        return keepsakesChosen as KeepsakeConfiguration;
    }

    maxLevelAll() {
        this.items.forEach(keepsake => keepsake.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(keepsake => keepsake.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(keepsake => keepsake.isMaxLevel);
    }

    fromStoredData(storedData: StorableKeepsakesData) {
        super.fromStoredData(storedData);
    }
}

interface StorableKeepsakeData extends StorableItemData {
    level: number;
}

export class Keepsake extends Item {
    level: Level = new Level(1, Keepsake.minLevel, Keepsake.maxLevel);
    constructor(public name: string, public giver: string, public icon: string, public isHidden: boolean = false) {
        super();
    }

    static minLevel = 1;
    static maxLevel = 3;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }

    toStorableData(): StorableKeepsakeData {
        const itemData = super.toStorableData();
        return {...itemData, level: this.level.value};
    }

    fromStoredData(storableData: StorableKeepsakeData) {
        super.fromStoredData(storableData);
        this.level.value = storableData.level;
    }
}

export type KeepsakeConfiguration = [Keepsake] | [Keepsake, Keepsake | null, Keepsake | null, Keepsake | null] | null;