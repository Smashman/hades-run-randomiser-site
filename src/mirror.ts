import { getRandomItemFromArray, Item, Options, StorableItemData } from './utils';

export interface MirrorOptions extends Options {
};

export const defaultMirrorOptions: MirrorOptions = {
    enabled: true,
};

export class Mirror {
    groups: MirrorGroup[] = [];

    addGroup(sets: [[MirrorTalent, MirrorTalent], [MirrorTalent, MirrorTalent]], isUnlocked: boolean = false) {
        const mirrorTalentSets = <[MirrorTalentSet, MirrorTalentSet]>sets.map(set => new MirrorTalentSet(set));
        this.groups = [...this.groups, new MirrorGroup(mirrorTalentSets, isUnlocked)];
        return this;
    }

    get unlockedGroups() {
        return this.groups.filter(group => group.isUnlocked);
    }

    unlockAll() {
        this.groups.forEach(group => group.unlock());
    }

    getRandom(): MirrorConfiguration {
        return new MirrorConfiguration(this.unlockedGroups.map(group => group.getRandom()).reduce((acc, val) => [...acc, ...val], <MirrorTalent[]>[]));
    }

    toStorableData() {
        return {
            groups: this.groups.map(group => group.toStorableData())
        }
    }
}

export class MirrorGroup extends Item{
    constructor(public sets: [MirrorTalentSet, MirrorTalentSet], public isUnlocked: boolean = false) {
        super();
    }

    getRandom() {
        return <[MirrorTalent, MirrorTalent]>this.sets.map(set => set.getRandom());
    }
}

export class MirrorTalentSet {
    constructor(public talents: [MirrorTalent, MirrorTalent]) {
        talents[0].colour = "red";
        talents[1].colour = "green";
    }

    getRandom(): MirrorTalent {
        return getRandomItemFromArray(this.talents);
    }
}

export class MirrorTalent {
    rankCount: number;
    colour: "red" | "green";
    totalCost: number;

    constructor(public name: string, public rankCosts: number[]) {
        this.rankCount = rankCosts.length;
        this.totalCost = rankCosts.reduce((totalCost, cost) => totalCost + cost);
    }
}

export class MirrorConfiguration {
    totalDarkness: number;
    constructor(public talents: MirrorTalent[]){
        this.totalDarkness = talents.reduce((totalCost, talent) => totalCost + talent.totalCost, 0);
    };
}