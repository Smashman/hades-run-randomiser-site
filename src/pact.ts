import { getRandomItemFromArray, Options } from './utils';

export interface HeatRange {
    min?: number;
    max?: number;
}

export interface PactOptions extends Options {
    heatRange?: HeatRange
}

export class Pact {
    conditions: PactCondition[] = [];

    addCondition(pactCondition: PactCondition) {
        this.conditions = [...this.conditions, pactCondition];
        return this;
    }

    getRandomWithRange(heatRange?: HeatRange) {
        return PactConfiguration.generateFromHeatRange(this, heatRange);
    }

    unlockAll() {
        this.conditions.forEach(condition => condition.enforceLimit = false);
    }

    get maxHeat(): number {
        return this.conditions.reduce((maxTotal, condition) => maxTotal + condition.totalHeat, 0);
    }

    get initialHeat(): number {
        return this.conditions.reduce((total, condition) => total + condition.initialHeat, 0);
    }
}

export class PactCondition {
    enforceLimit: boolean = false;
    private _initialRank: number = 0;
    constructor(public name: string, public readonly rankCosts: number[], public rankLimit?: number) {
        this.enforceLimit = !!rankLimit;
    }

    get initialRank(): number {
        return this._initialRank;
    }

    set initialRank(rank: number) {
        this._initialRank = rank > this.rankCount ? this.rankCount : rank;
    }

    get rankCount(): number {
        return this.rankLimit && this.enforceLimit ? this.rankLimit : this.rankCosts.length;
    }

    get totalHeat(): number {
        return (this.rankLimit && this.enforceLimit ? this.rankCosts.slice(0, this.rankLimit) : this.rankCosts).reduce((totalHeat, cost) => totalHeat + cost);
    }

    get initialHeat(): number {
        return this.rankCosts.slice(0, this.initialRank).reduce((total, value) => total + value, 0);
    }
}

export class PactConfiguration {
    conditionConfigurations: PactConditionConfiguration[];
    heatTotal: number = 0;
    heatTarget: number = 0;
    heatRange: HeatRange;

    constructor(pact: Pact, heatRange?: HeatRange, ranks?: number[]) {
        this.conditionConfigurations = pact.conditions.map((condition, index) => new PactConditionConfiguration(condition, (ranks && ranks[index]) || null));
        this.heatRange = {
            min: heatRange?.min || pact.initialHeat,
            max: heatRange?.max || pact.maxHeat,
        };
    }

    static generateFromHeatRange(pact: Pact, heatRange?: HeatRange) {
        const pactConfiguration = new PactConfiguration(pact, heatRange);

        const configHeatRange = pactConfiguration.heatRange;


        const target = Math.floor(Math.random() * (configHeatRange.max - configHeatRange.min + 1) + configHeatRange.min);

        let current = pact.initialHeat;

        let remainingConditions = [...pactConfiguration.conditionConfigurations.filter((conditionConfig) => conditionConfig.rank !== conditionConfig.condition.rankCount)];


        while (current < target) {
            const randomCondition = getRandomItemFromArray(remainingConditions);
            current += randomCondition.rankUp();

            if (randomCondition.rank === randomCondition.condition.rankCount) {
                remainingConditions = [...remainingConditions.filter(condition => randomCondition !== condition)];
            }
        }

        pactConfiguration.heatTotal = current;
        pactConfiguration.heatTarget = target;

        return pactConfiguration;
    }
}

export class PactConditionConfiguration {
    private _rank: number = 0;

    constructor(public condition: PactCondition, rank?: number) {
        this.rank = rank || condition.initialRank;
    }

    get heat(): number {
        return this.condition.rankCosts.slice(0, this.rank).reduce((total, value) => total + value, 0);
    }

    get rank(): number {
        return this._rank;
    }

    set rank(rank: number) {
        this._rank = rank > this.condition.rankCount ? this.condition.rankCount : rank;
    }

    rankUp(): number {
        if (this.rank < this.condition.rankCount) {
            this.rank++;
            return this.condition.rankCosts[this.rank-1];
        }
        else {
            return null;
        }
    }
}