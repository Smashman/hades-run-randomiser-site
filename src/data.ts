import { Weapon, WeaponAspect, Weapons } from './weapons';
import { Keepsake, Keepsakes } from './keepsakes';
import { Companion, Companions } from './companions';
import { Mirror, MirrorTalent } from './mirror';
import { Pact, PactCondition } from './pact';
import { fillArray } from './utils';
import * as React from 'react';
import * as keepsakeIcons from './img/keepsake';
import * as weaponIcons from './img/weapon';

export const weapons = new Weapons().addItems([
    new Weapon('Stygius', 'Blade of the Underworld', 'sword', true)
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.sword.zagreus, false, true),
        new WeaponAspect('Nemesis', weaponIcons.sword.nemesis),
        new WeaponAspect('Poseidon', weaponIcons.sword.poseidon),
        new WeaponAspect('Arthur', weaponIcons.sword.arthur, true),
    ]),
    new Weapon('Varatha', 'Eternal Spear', 'spear')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.spear.zagreus, false, true),
        new WeaponAspect('Achilles', weaponIcons.spear.achilles),
        new WeaponAspect('Hades', weaponIcons.spear.hades),
        new WeaponAspect('Guan Yu', weaponIcons.spear.guanYu, true),
    ]),
    new Weapon('Aegis', 'Shield of Chaos', 'shield')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.shield.zagreus, false, true),
        new WeaponAspect('Chaos', weaponIcons.shield.chaos),
        new WeaponAspect('Zeus', weaponIcons.shield.zeus),
        new WeaponAspect('Beowulf', weaponIcons.shield.beowulf, true),
    ]),
    new Weapon('Coronacht', 'Heart-Seeking Bow', 'bow')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.bow.zagreus, false, true),
        new WeaponAspect('Chiron', weaponIcons.bow.chiron),
        new WeaponAspect('Hera', weaponIcons.bow.hera),
        new WeaponAspect('Rama', weaponIcons.bow.rama, true),
    ]),
    new Weapon('Malphon', 'Twin Fists', 'fists')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.fists.zagreus, false, true),
        new WeaponAspect('Talos', weaponIcons.fists.talos),
        new WeaponAspect('Demeter', weaponIcons.fists.demeter),
        new WeaponAspect('Gilgamesh', weaponIcons.fists.gilgamesh, true),
    ]),
    new Weapon('Exagryph', 'Adamant Rail', 'gun')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.gun.zagreus, false, true),
        new WeaponAspect('Eris', weaponIcons.gun.eris),
        new WeaponAspect('Hestia', weaponIcons.gun.hestia),
        new WeaponAspect('Lucifer', weaponIcons.gun.lucifer, true),
    ]),
]);

export const keepsakes = new Keepsakes().addItems([
    new Keepsake('Old Spiked Collar', 'Cerberus', keepsakeIcons.oldSpikedCollar),
    new Keepsake('Myrmidon Bracer', 'Achilles', keepsakeIcons.myrmidonBracer),
    new Keepsake('Black Shawl', 'Nyx', keepsakeIcons.blackShawl),
    new Keepsake('Pierced Butterfly', 'Thanatos', keepsakeIcons.piercedButterfly),
    new Keepsake('Bone Hourglass', 'Charon', keepsakeIcons.boneHourglass),
    new Keepsake('Chthonic Coin Purse', 'Hypnos', keepsakeIcons.chthonicCoinPurse),
    new Keepsake('Skull Earring', 'Megaera', keepsakeIcons.skullEarring),
    new Keepsake('Distant Memory', 'Orpheus', keepsakeIcons.distantMemory),
    new Keepsake('Harpy Feather Duster', 'Dusa', keepsakeIcons.harpyFeatherDuster),
    new Keepsake('Lucky Tooth', 'Skelly', keepsakeIcons.luckyTooth),
    new Keepsake('Thunder Signet', 'Zeus', keepsakeIcons.thunderSignet),
    new Keepsake('Conch Shell', 'Poseidon', keepsakeIcons.conchShell),
    new Keepsake('Owl Pendant', 'Athena', keepsakeIcons.owlPendant),
    new Keepsake('Eternal Rose', 'Aphrodite', keepsakeIcons.eternalRose),
    new Keepsake('Blood-Filled Vial', 'Ares', keepsakeIcons.bloodFilledVial),
    new Keepsake('Adamant Arrowhead', 'Artemis', keepsakeIcons.adamantArrowhead),
    new Keepsake('Overflowing Cup', 'Dionysus', keepsakeIcons.overflowingCup),
    new Keepsake('Lambent Plume', 'Hermes', keepsakeIcons.lambentPlume),
    new Keepsake('Frostbitten Horn', 'Demeter', keepsakeIcons.frostbittenHorn),
    new Keepsake('Cosmic Egg', 'Chaos', keepsakeIcons.cosmicEgg),
    new Keepsake('Shattered Shackle', 'Sisyphus', keepsakeIcons.shatteredShackle),
    new Keepsake('Evergreen Acorn', 'Eurydice', keepsakeIcons.evergreenAcorn),
    new Keepsake('Broken Spearpoint', 'Patroclus', keepsakeIcons.brokenSpearpoint),
    new Keepsake('Pom Blossom', 'Persephone', keepsakeIcons.pomBlossom),
    new Keepsake('Sigil of the Dead', 'Hades', keepsakeIcons.sigilOfTheDead, true),
]);

export const companions = new Companions().addItems([
    new Companion('Battie', 'Megaera'),
    new Companion('Mort', 'Thanatos'),
    new Companion('Rib', 'Skelly'),
    new Companion('Shady', 'Sisyphus'),
    new Companion('Fidi', 'Dusa'),
    new Companion('Antos', 'Achilles'),
]);

export const mirror = new Mirror()
    .addGroup([
        [
            new MirrorTalent('Shadow Presence', [10, 15, 20, 25, 30]),
            new MirrorTalent('Fiery Presence', [10, 15, 20, 25, 30]),
        ], [
            new MirrorTalent('Chthonic Vitality', [10, 20, 40]),
            new MirrorTalent('Dark Regeneration', [30, 60]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Death Defiance', [30, 500, 1000]),
            new MirrorTalent('Stubborn Defiance', [600]),
        ],
        [
            new MirrorTalent('Greater Reflex', [50]),
            new MirrorTalent('Ruthless Reflex', [75]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Boiling Blood', [10, 30, 50, 70, 90]),
            new MirrorTalent('Abyssal Blood', [20, 40, 60, 80, 100]),
        ],
        [
            new MirrorTalent('Infernal Soul', [20, 80]),
            new MirrorTalent('Stygian Soul', [0, 60, 180]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Deep Pockets', [30, 35, 40, 45, 50, 55, 60, 65, 70, 75]),
            new MirrorTalent('Golden Touch', [70, 90, 110]),
        ],
        [
            new MirrorTalent('Thick Skin', [40, 45, 50, 55, 60, 65, 70, 75, 80, 85]),
            new MirrorTalent('High Confidence', [50, 100, 150, 200, 250]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Privileged Status', [50, 500]),
            new MirrorTalent('Family Favorite', [50, 150]),
        ],
        [
            new MirrorTalent('Olympian Favor', fillArray(40, 50)),
            new MirrorTalent('Dark Foresight', fillArray(10, 150)),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Gods\' Pride', fillArray(20, 100)),
            new MirrorTalent('Gods\' Legacy', fillArray(10, 250)),
        ],
        [
            new MirrorTalent('Fated Authority', [500, 750, 1000, 1250, 1500, 1750, 2000, 2250]),
            new MirrorTalent('Fated Persuasion', [1000, 2000, 3000, 4000]),
        ]
    ]);

export const pact = new Pact()
    .addCondition(new PactCondition('Hard Labor', [1, 1, 1, 1, 1]))
    .addCondition(new PactCondition('Lasting Consequences', [1, 1, 1, 1]))
    .addCondition(new PactCondition('Convenience Fee', [1, 1]))
    .addCondition(new PactCondition('Jury Summons', [1, 1, 1]))
    .addCondition(new PactCondition('Extreme Measures', [1, 2, 3, 4], 3))
    .addCondition(new PactCondition('Calisthenics Program', [1, 1]))
    .addCondition(new PactCondition('Benefits Package', [2, 3]))
    .addCondition(new PactCondition('Middle Management', [2]))
    .addCondition(new PactCondition('Underworld Customs', [2]))
    .addCondition(new PactCondition('Forced Overtime', [3, 3]))
    .addCondition(new PactCondition('Heightened Security', [1]))
    .addCondition(new PactCondition('Routine Inspection', [2, 2, 2, 2]))
    .addCondition(new PactCondition('Damage Control', [1, 1]))
    .addCondition(new PactCondition('Approval Process', [2, 3]))
    .addCondition(new PactCondition('Tight Deadline', [1, 2, 3]));

// weapons.unlockAll();
// keepsakes.unlockAll();
// companions.unlockAll();
// mirror.unlockAll();
// pact.unlockAll();

// weapons.maxLevelAll();
// keepsakes.maxLevelAll();
// companions.maxLevelAll();
// companions.maxCodexAll();

interface Data {
    weapons: Weapons;
    keepsakes: Keepsakes;
    companions: Companions;
    mirror: Mirror;
    pact: Pact;
}

export const defaultData: Data = {
    weapons,
    keepsakes,
    companions,
    mirror,
    pact,
};

export const DataContext = React.createContext<[Data, React.Dispatch<React.SetStateAction<Data>>]>([defaultData, {} as React.Dispatch<React.SetStateAction<Data>>]);

//@ts-ignore
// window.data = defaultData;