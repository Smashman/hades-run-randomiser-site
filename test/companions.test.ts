import { Companion, Companions } from '../src/companions';
import * as utils from '../src/utils';
import { getRandomItemFromArrayMock } from './helpers';

describe('Companion class', () => {
    it('should create expected Companion class instance', () => {
        const companion = new Companion('Fluff', 'Jeff');

        expect(companion).toMatchSnapshot();
    });

    it('should return false for isMaxLevel when level is less than 5', () => {
        const companion = new Companion('Fluff', 'Jeff');
        companion.level.value = 4;

        expect(companion.isMaxLevel).toBe(false);
    });

    it('should return true for isMaxLevel when level is 5', () => {
        const companion = new Companion('Fluff', 'Jeff');
        companion.level.value = 5;

        expect(companion.isMaxLevel).toBe(true);
    });

    it('should change resource values when level is updated', () => {
        const companion = new Companion('Fluff', 'Jeff');

        expect(companion.resource.total).toBe(10);
        expect(companion.resource.spent).toBe(0);
        expect(companion.resource.required).toBe(10);

        companion.level.value = 4;

        expect(companion.resource.total).toBe(10);
        expect(companion.resource.spent).toBe(6);
        expect(companion.resource.required).toBe(4);
    });
});

describe('Companions class', () => {
    it('should return false for isAllMaxLevel if not every companion is max level', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companion1.level.value = 5;

        expect(companions.isAllMaxLevel).toBe(false);
    });

    it('should return true for isAllMaxLevel if every companion is max level', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companion1.level.value = 5;
        companion2.level.value = 5;

        expect(companions.isAllMaxLevel).toBe(true);
    });

    it('should max level all companions when maxLevelAll is called', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companions.maxLevelAll();

        expect(companion1.level.value).toBe(5);
        expect(companion2.level.value).toBe(5);

        expect(companions.isAllMaxLevel).toBe(true);
    });

    it('should return false for isAllMaxCodex if not every companion is max codex', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companion1.isMaxCodex = true;

        expect(companions.isAllMaxCodex).toBe(false);
    });

    it('should return true for isAllMaxCodex if every companion is max codex', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companion1.isMaxCodex = true;
        companion2.isMaxCodex = true;

        expect(companions.isAllMaxCodex).toBe(true);
    });

    it('should max codex all companions when maxCodexAll is called', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companions.maxCodexAll();

        expect(companion1.isMaxCodex).toBe(true);
        expect(companion2.isMaxCodex).toBe(true);

        expect(companions.isAllMaxCodex).toBe(true);
    });

    it('should calculate resource values from all companions', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        expect(companions.resourceTotal).toBe(20);
        expect(companions.resourceSpent).toBe(0);
        expect(companions.resourceRequired).toBe(20);
    });

    it('should calculate resource values from all companions when aspects are leveled', () => {
        const companion1 = new Companion('Fluff', 'Jeff');
        const companion2 = new Companion('Rex', 'Delilah');
        const companions = new Companions().addItems([companion1, companion2]);

        companion1.level.value = 2;
        companion2.level.value = 3;

        expect(companions.resourceTotal).toBe(20);
        expect(companions.resourceSpent).toBe(4);
        expect(companions.resourceRequired).toBe(16);
    });

    describe('getRandomCompanion', () => {
        let getRandomItemFromArraySpy: jest.SpyInstance;

        let companions: Companions;

        let companion1: Companion;
        let companion2: Companion;
        let companion3: Companion;
        let companion4: Companion;

        beforeEach(() => {
            companion1 = new Companion('Fluff', 'Jeff');
            companion2 = new Companion('Rex', 'Delilah');
            companion3 = new Companion('Trunky', 'Bob');
            companion4 = new Companion('Mr. Bilbadopulous', 'Trevor');
            companions = new Companions().addItems([companion1, companion2, companion3, companion4]);

            getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);
        });

        afterEach(() => {
            jest.spyOn(utils, 'getRandomItemFromArray').mockRestore();
        });

        it('should return null if no companions are unlocked', () => {
            const randomCompanion = companions.getRandomCompanion();

            expect(getRandomItemFromArraySpy).toBeCalled();
            expect(randomCompanion).toBe(null);
        });

        it('should return expected companion if all companions are unlocked', () => {
            companions.unlockAll();

            const randomCompanion = companions.getRandomCompanion();

            expect(getRandomItemFromArraySpy).toBeCalledWith([companion1, companion2, companion3, companion4]);
            expect(randomCompanion).toBe(companion3);
        });

        it('should return expected companion when a subset of companions are unlocked', () => {
            companion2.unlock();
            companion4.unlock();

            const randomCompanion = companions.getRandomCompanion();

            expect(getRandomItemFromArraySpy).toBeCalledWith([companion2, companion4]);
            expect(randomCompanion).toBe(companion4);
        });

        it('should return expected companion with ignoreMaxCodex option but none are maxCodex', () => {
            companions.unlockAll();

            const randomCompanion = companions.getRandomCompanion({ignoreMaxCodex: true});

            expect(getRandomItemFromArraySpy).toBeCalledWith([companion1, companion2, companion3, companion4]);
            expect(randomCompanion).toBe(companion3);
        });

        it('should return expected companion with ignoreMaxCodex option and one is maxCodex', () => {
            companions.unlockAll();

            companion3.isMaxCodex = true;

            const randomCompanion = companions.getRandomCompanion({ignoreMaxCodex: true});

            expect(getRandomItemFromArraySpy).toBeCalledWith([companion1, companion2, companion4]);
            expect(randomCompanion).toBe(companion2);
        });
    });

    describe('toStorableData', () => {
        it('should return expected storable data', () => {
            const companion = new Companion('Fluff', 'Jeff');

            expect(companion.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with level', () => {
            const companion = new Companion('Fluff', 'Jeff');
            companion.level.value = 3;

            expect(companion.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with isMaxCodex', () => {
            const companion = new Companion('Fluff', 'Jeff');
            companion.isMaxCodex = true;

            expect(companion.toStorableData()).toMatchSnapshot();
        });
    });
});