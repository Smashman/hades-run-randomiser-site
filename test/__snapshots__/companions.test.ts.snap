// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Companion class should create expected Companion class instance 1`] = `
Companion {
  "giver": "Jeff",
  "isMaxCodex": false,
  "isUnlocked": false,
  "level": Level {
    "_value": 1,
    "max": 5,
    "min": 1,
    "onChange": undefined,
  },
  "name": "Fluff",
}
`;

exports[`Companions class toStorableData should return expected storable data 1`] = `
Object {
  "isMaxCodex": false,
  "isUnlocked": false,
  "level": 1,
}
`;

exports[`Companions class toStorableData should return expected storable data with isMaxCodex 1`] = `
Object {
  "isMaxCodex": true,
  "isUnlocked": false,
  "level": 1,
}
`;

exports[`Companions class toStorableData should return expected storable data with level 1`] = `
Object {
  "isMaxCodex": false,
  "isUnlocked": false,
  "level": 3,
}
`;
